lib/core/src/Cardano/Wallet/Primitive/Types/TokenMap.hs

```diff
@@ -0,0 +1,539 @@
+    parseJSON = genericParseJSON jsonOptions
+instance ToJSON NestedMapEntry where
+    toJSON = genericToJSON jsonOptions
+
+instance FromJSON NestedTokenQuantity where
+    parseJSON = genericParseJSON jsonOptions
+instance ToJSON NestedTokenQuantity where
+    toJSON = genericToJSON jsonOptions
+
+--------------------------------------------------------------------------------
+-- Construction
+--------------------------------------------------------------------------------
+
+-- | The empty token map.
+--
+empty :: TokenMap
+empty = TokenMap mempty
+
+-- | Creates a singleton token map with just one token quantity.
+--
+-- If the specified token quantity is zero, then the resultant map will be
+-- equal to the 'empty' map.
+--
+singleton :: AssetId -> TokenQuantity -> TokenMap
+singleton = setQuantity empty
+
+-- | Creates a token map from a flat list.
+--
+-- If a token name appears more than once in the list under the same policy,
+-- its associated quantities will be added together in the resultant map.
+--
+fromFlatList :: [(AssetId, TokenQuantity)] -> TokenMap
+fromFlatList = F.foldl' acc empty
+  where
+    acc b (asset, quantity) = adjustQuantity b asset (<> quantity)
+
+-- | Creates a token map from a nested list.
+--
+-- If a token name appears more than once in the list under the same policy,
+-- its associated quantities will be added together in the resultant map.
+--
+fromNestedList
+    :: [(TokenPolicyId, NonEmpty (TokenName, TokenQuantity))] -> TokenMap
+fromNestedList entries = fromFlatList
+    [ (AssetId policy token, quantity)
+    | (policy, tokenQuantities) <- entries
+    , (token, quantity) <- NE.toList tokenQuantities
+    ]
+
+--------------------------------------------------------------------------------
+-- Deconstruction
+--------------------------------------------------------------------------------
+
+-- | Converts a token map to a flat list.
+--
+toFlatList :: TokenMap -> [(AssetId, TokenQuantity)]
+toFlatList b =
+    [ (AssetId policy token, quantity)
+    | (policy, tokenQuantities) <- toNestedList b
+    , (token, quantity) <- NE.toList tokenQuantities
+    ]
+
+-- | Converts a token map to a nested list.
+--
+toNestedList
+    :: TokenMap -> [(TokenPolicyId, NonEmpty (TokenName, TokenQuantity))]
+toNestedList =
+    fmap (fmap NEMap.toList) . Map.toList . unTokenMap
+
+--------------------------------------------------------------------------------
+-- Arithmetic
+--------------------------------------------------------------------------------
+
+-- | Adds one token map to another.
+--
+add :: TokenMap -> TokenMap -> TokenMap
+add a b = F.foldl' acc a $ toFlatList b
+  where
+    acc c (asset, quantity) =
+        adjustQuantity c asset (`TQ.add` quantity)
+
+-- | Subtracts one token map from another.
+--
+subtract :: TokenMap -> TokenMap -> TokenMap
+subtract a b = F.foldl' acc a $ toFlatList b
+  where
+    acc c (asset, quantity) =
+        adjustQuantity c asset (`TQ.subtract` quantity)
```

Comment:
After reviewing with @rvl, we decided to remove the `subtract` operation for the moment, as it's not clear that it gives us anything of value.

If we need to add this back in later, we can.

Reply:
Fixed in 5bbd9b34a13a5b5d0b517fd9b2d19e3a1b9d73df.